insights/parsers/satellite_postgresql_query.py

```diff
@@ -0,0 +1,177 @@
+                    data.update(dict(zip(self._columns[1:], [[] for i in range(1, len(self._columns))])))
+                    self.append(data)
+                    for i in range(1, len(self._columns)):
+                        if row_data[i]:
+                            data[self._columns[i]].append(row_data[i])
+                else:
+                    for i in range(1, len(self._columns)):
+                        if row_data[i]:
+                            data[self._columns[i]].append(row_data[i])
+        if not len(self):
+            raise SkipException("There is no data")
+        if self.count != len(self):
+            raise ParseException('The count of rows is not equal to the number of the last row')
+
+    def __contains__(self, first_column_value):
+        return True if any(
+            row[self._first_columm] == first_column_value for row in self
+        ) else False
+
+    def get_column_value(self, first_column_value, column_name):
+        '''
+        Returns:
+            An empty_string when there is no value,
+            the first element when the value has only one line,
+            the value itself when the value has multiple lines
+
+        Args:
+            first_column_value(str): the value of the first column to identify the row
+            column_name(str): the colunm name you want to get the value for
+        '''
+
+        if column_name not in self._columns:
+            raise ParseException("No column name %s found" % column_name)
+        for row in self:
+            if row[self._first_columm] == first_column_value:
+                if isinstance(row[column_name], list):
+                    if len(row[column_name]) == 0:
+                        return ''
+                    if len(row[column_name]) == 1:
+                        return row[column_name][0]
+                return row[column_name]
+        raise ParseException("No column value %s found" % first_column_value)
+
+
+@parser(Specs.satellite_settings)
+class SatelliteAdminSettings(SatellitePostgreSQLQuery):
+    """
+    Parse the output of the command ``psql -d foreman -c '"select name, value, "default" from settings where name in ('destroy_vm_on_host_delete', 'unregister_delete_host')"``
+    and save the data to itself as a dict.
+
+    Sample output::
+
+                name               |  value   |  default
+        ---------------------------+----------+-----------
+        unregister_delete_host     | --- true+| --- false+
+                                   | ...      | ...
+        destroy_vm_on_host_delete  |          | --- true +
+                                   |          | ...
+        (2 rows)
+
+    Examples::
+
+        >>> type(settings)
+        <class 'insights.parsers.satellite_postgresql_query.SatelliteAdminSettings'>
+        >>> settings.get_settting('unregister_delete_host')
+        True
+        >>> len(settings)
+        2
+        >>> settings.get_settting('destroy_vm_on_host_delete')
```

Comment:
@bfahr, in the base class, I'm not sure if the value of one column is unique for all the tables, so I save the rows to a list.
  Although in the settings table, the name is actually unique,  in the rule I just want to get the value of some setting, I don't override  the format by looping it again after extending the base class. Do u think we have to override it?

Reply:
@huali027 there's 3 issues with `get_settings`.  First are there only 2 settings that a user would pass to this method, those in the documentation? If not then how would a user get all of the setting names to know what to pass to the method?

Second, you looping every time this method is called.  You could create a `dict` with all valid settings names and then just do a lookup in the dict each time the method is called.

Third, you are checking each time to determine if both `valid` and `default` columns are in the row and then raising an `ParseException` if they are not.  This should only be checked in `parse_content`, and not in the `get_settings` method.  So if you created the `dict` in `parse_content` of all settings then you could verify at that time whether or not the data is valid.  You don't want to have rules checking for ParseExceptions.

Reply:
@bfahr,
For the first issue, yes, since my rule only uses these two settings, I limit the SQL to only select these two settings. There are about 240 records in the table. To save time, I add a limitation. Because maybe the other settings are never used.
The rule user should know which settings he wants to use. So we don't need to provide a method to query all the settings.
Sure, I'll update the code to save settings table into a dict format.